### PR TITLE
Async runtime

### PR DESCRIPTION
**Pull request by Gavin-Niederman**
Add a basic async runtime to pros-rs and make Robot an async trait.

todo before merging:
* [ ] Test the async runtime on real vex hardware

### CODE REVIEW

File: pros/examples/accessories.rs [hunk 1]
```diff
@@ -6,10 +6,21 @@ use pros::prelude::*;
 
 #[derive(Debug, Default)]
 struct ExampleRobot;
+#[async_trait]
```
Comment (doinkythederp) [inline pros/examples/accessories.rs:9]: Thoughts on waiting for async traits to stabilize?

File: pros/src/async_runtime/executor.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    task::{Context, Waker},
+};
+
+use alloc::{boxed::Box, collections::VecDeque, rc::Rc, task::Wake};
+
+use crate::task_local;
+
+use super::{reactor::Reactor, JoinHandle};
+
+task_local! {
```
Comment (doinkythederp) [inline pros/src/async_runtime/executor.rs:14]: I feel like task_local is a bit of a confusing name because we have both freertos tasks and async tasks.

File: pros/src/async_runtime/executor.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    pub(crate) static EXECUTOR: Rc<Executor> = Rc::new(Executor::new())
```
Comment (doinkythederp) [inline pros/src/async_runtime/executor.rs:15]: Does this mean async can't be used between multiple threads?

File: pros/src/async_runtime/executor.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        loop {
+            self.tick();
+
+            if let Some(output) = (*output).borrow_mut().take() {
+                break output;
+            }
+        }
+    }
+
+    pub fn complete(&self) {
```
Comment (doinkythederp) [inline pros/src/async_runtime/executor.rs:102]: Imo it would nice if this had a version with a timeout

File: pros/src/async_runtime/executor.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        while self.tick() {}
+    }
+}
+
+pub struct TaskWaker {
+    task: RefCell<Option<Task>>,
+}
+// These are here to apease the waker struct.
+// The executor is single threaded and this waker will never be passed around threads or shared between threads.
+unsafe impl Send for TaskWaker {}
```
Comment (doinkythederp) [inline pros/src/async_runtime/executor.rs:112]: If they will never be sent between threads, why is this necessary?

File: pros/src/async_runtime/mod.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
+use core::{cell::RefCell, future::Future};
+
+use alloc::rc::Rc;
+
+pub(crate) mod executor;
+pub(crate) mod reactor;
+
+pub struct JoinHandle<T> {
+    output: Rc<RefCell<Option<T>>>,
+    _marker: core::marker::PhantomData<T>,
```
Comment (doinkythederp) [inline pros/src/async_runtime/mod.rs:10]: I don't think the marker field is necessary here.

File: pros/src/async_runtime/mod.rs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    executor::EXECUTOR.with(|e| e.spawn(future))
+}
+
+pub fn block_on<F: Future + 'static>(future: F) -> F::Output {
+    executor::EXECUTOR.with(|e| e.block_on(future))
+}
+
+pub fn complete() {
+    executor::EXECUTOR.with(|e| e.complete());
+}
```
Comment (doinkythederp) [inline pros/src/async_runtime/mod.rs:41]: Could these get docs? Even just a super simple explanation of what they do would be great

File: pros/src/motor.rs [hunk 8]
```diff
@@ -274,6 +279,25 @@ impl From<i32> for Gearset {
...
+        self: core::pin::Pin<&mut Self>,
+        cx: &mut core::task::Context<'_>,
+    ) -> core::task::Poll<Self::Output> {
+        match self.motor.get_state()?.stopped {
+            true => core::task::Poll::Ready(Ok(())),
+            false => {
+                cx.waker().wake_by_ref();
+                core::task::Poll::Pending
+            }
+        }
```
Comment (doinkythederp) [inline pros/src/motor.rs:297]: This seems just like an if statement but with more steps

**Review by doinkythederp — CHANGES_REQUESTED** [review]
Some thoughts and questions I had

File: pros/src/async_runtime/executor.rs (hunk 2 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/executor.rs:14]: Async tasks are only used internally, also the std feature this is mimicking is reference to the equivalent of freertos tasks. However I do feel it is worth discussing whether we should change the name of task to thread just in general in another pr.

File: pros/examples/accessories.rs (hunk 1 above)
Comment (Gavin-Niederman) [reply pros/examples/accessories.rs:9]: I was thinking about that, but it might be a while until then. The only real upside to waiting is that we will have a slightly better API on release, but it is better, in my opinion, to get async out earlier than later because of how much it improves things. It would also not be a very hard change to make later.

File: pros/src/async_runtime/executor.rs (hunk 2 above)
Comment (doinkythederp) [reply pros/src/async_runtime/executor.rs:14]: Maybe this could be called `os_task_local` for now?

File: pros/src/async_runtime/executor.rs (hunk 3 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/executor.rs:15]: I'm not entirely sure what you mean by this. If it helps, every freertos task has its own executor because the executor is single threaded. If tasks shared the same executor a future being polled in one task would block a future being polled in a completely different one, which could dramatically slow things down in certain circumstances.

File: pros/src/async_runtime/executor.rs (hunk 5 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/executor.rs:112]: It is necessary because the Wake trait in alloc requires implementors to implement Send and Sync because most async executors are multi-threaded. TaskWaker isn't Send and Sync because Task isn't either.

File: pros/src/async_runtime/mod.rs (hunk 6 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/mod.rs:10]: It isn't that should be fixed.

**Review by Gavin-Niederman — COMMENTED** [review]
remove unused marker

File: pros/src/motor.rs (hunk 8 above)
Comment (Gavin-Niederman) [reply pros/src/motor.rs:297]: you could also write it with an if, but a match works just fine. I specifically chose a match here because I thought it looked better.

File: pros/src/async_runtime/executor.rs (hunk 4 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/executor.rs:102]: Do you mean a function that polls tasks until a certain amount of time has passed?

File: pros/src/async_runtime/executor.rs (hunk 2 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/executor.rs:14]: That would work.

File: pros/src/async_runtime/mod.rs (hunk 7 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/mod.rs:41]: that would be a good addition.

File: pros/src/async_runtime/executor.rs (hunk 4 above)
Comment (doinkythederp) [reply pros/src/async_runtime/executor.rs:102]: Yeah, I feel like that could be used for like "wait a few seconds for everything to finish up, but if it takes too long, just bail out" or even to check if everything is finished by passing in 0.

File: pros/src/async_runtime/mod.rs (hunk 7 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/mod.rs]: Documented

File: pros/src/async_runtime/mod.rs (hunk 6 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/mod.rs]: done

File: pros/src/async_runtime/executor.rs (hunk 3 above)
Comment (doinkythederp) [reply pros/src/async_runtime/executor.rs:15]: What I meant is that if I were to make a future and send it to another thread, could I use it there? I'm guessing yes but I'm not positive.

Comment (Gavin-Niederman) [reply pros/src/async_runtime/executor.rs:15]: If you register a future in the executor there is no way to send it. The executor owns futures. This unsafe impl is on the TaskWaker struct which only is ever created in an executor. That is why it is fine to unsafe impl.

File: pros/examples/accessories.rs [hunk 9]
```diff
@@ -6,10 +6,21 @@ use pros::prelude::*;
...
 struct ExampleRobot;
-impl Robot for ExampleRobot {
-    fn opcontrol(&mut self) -> pros::Result {
+#[async_trait]
+impl AsyncRobot for ExampleRobot {
+    async fn opcontrol(&mut self) -> pros::Result {
+        let handle = pros::async_runtime::spawn(async {
+            for _ in 0..5 {
+                println!("Hello from async!");
+                sleep(Duration::from_millis(1000));
```
Comment (doinkythederp) [inline pros/examples/accessories.rs:15]: Should this be awaited?

File: pros/examples/basic.rs [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+#![no_main]
+
+use pros::prelude::*;
+
+#[derive(Default)]
+pub struct Robot;
+#[async_trait]
+impl AsyncRobot for Robot {
+    async fn opcontrol(&mut self) -> pros::Result {
+        println!("basic exasmple");
```
Comment (doinkythederp) [inline pros/examples/basic.rs:11]: Typo here

File: pros/src/async_runtime/executor.rs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        Self {
+            queue: RefCell::new(VecDeque::new()),
+            reactor: RefCell::new(Reactor::new()),
+        }
+    }
+
+    pub fn spawn<T>(&'static self, future: impl Future<Output = T> + 'static) -> Task<T> {
+        // SAFETY: `runnable` will never be moved off this thread or shared with another thread because of the `!Send + !Sync` bounds on `Self`.
+        //         Both `future` and `schedule` are `'static` so they cannot be used after being freed.
+        //   TODO: Make sure that the waker can never be sent off the thread.
```
Comment (doinkythederp) [inline pros/src/async_runtime/executor.rs:40]: I dont know if this needs more work but might be good to fix after merge

File: pros/src/async_runtime/mod.rs [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+/// Runs a future in the background without having to await it
+/// To get the the return value you can call [`JoinHandle.join`](JoinHandle::join).
+pub fn spawn<T>(future: impl Future<Output = T> + 'static) -> Task<T> {
+    executor::EXECUTOR.with(|e| e.spawn(future))
+}
+
+/// Blocks the current task untill a return value can be extracted from the provided future.
+/// Does not poll all futures to completion.
+/// If you want to complete all futures, use the [`complete`] function.
+pub fn block_on<F: Future + 'static>(future: F) -> F::Output {
```
Comment (doinkythederp) [inline pros/src/async_runtime/mod.rs:17]: where is the `complete` function?

**Review by doinkythederp — CHANGES_REQUESTED** [review]
Just some things that might be mistakes and then let's merge

File: pros/src/async_runtime/executor.rs (hunk 11 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/executor.rs:40]: This will be fixed after this is merged and I split sync and async functionality from each other. I will put them into separate features that are mutually exclusive so that this cant really ever be an issue.

File: pros/src/async_runtime/mod.rs (hunk 12 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/mod.rs:17]: A complete function would be useful. I will add one now.

File: pros/examples/basic.rs (hunk 10 above)
Comment (Gavin-Niederman) [reply pros/examples/basic.rs]: fixed

File: pros/examples/accessories.rs (hunk 9 above)
Comment (Gavin-Niederman) [reply pros/examples/accessories.rs]: fixed along with other mistakes in the same file.

File: pros/src/async_runtime/mod.rs (hunk 12 above)
Comment (Gavin-Niederman) [reply pros/src/async_runtime/mod.rs:17]: done

**Review by doinkythederp — APPROVED** [review]
LGTM (Lets Go To vex and ask them for Money) :rocket:
